src/main/java/org/cactoos/io/Zip.java

```diff
@@ -30,53 +30,69 @@
 import java.io.FileInputStream;
 import java.io.InputStream;
 import java.nio.file.Path;
+import java.util.Iterator;
 import java.util.zip.ZipEntry;
 import java.util.zip.ZipOutputStream;
 import org.cactoos.Input;
+import org.cactoos.func.ForEach;
+import org.cactoos.iterator.Skipped;
+import org.cactoos.text.Joined;
 
 /**
  * Zip files and directory.
  * <br>
  * <br>There is no thread-safety guarantee.
  *
  * @since 0.29
+ * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
  */
 @SuppressWarnings("PMD.AvoidInstantiatingObjectsInLoops")
 public final class Zip implements Input {
 
     /**
-     * Origin directory.
+     * Origin list of files.
      */
-    private final Directory origin;
+    private final Directory directory;
 
     /**
      * Ctor.
      *
-     * @param origin Origin directory.
+     * @param dir Origin directory.
      */
-    public Zip(final Directory origin) {
-        this.origin = origin;
+    public Zip(final Directory dir) {
+        this.directory = dir;
     }
 
     @Override
     public InputStream stream() throws Exception {
         final ByteArrayOutputStream out = new ByteArrayOutputStream();
         try (ZipOutputStream zip = new ZipOutputStream(out)) {
-            for (final Path path : this.origin) {
-                final File file = path.toFile();
-                final ZipEntry entry = new ZipEntry(
-                    file.getPath()
-                );
-                zip.putNextEntry(entry);
-                if (file.isFile()) {
-                    try (
-                        FileInputStream input = new FileInputStream(file)
-                    ) {
-                        zip.write(new BytesOf(new InputOf(input)).asBytes());
+            new ForEach<Path>(
+                path -> {
+                    final String relative = this.directory.path().getParent()
+                        .relativize(path).toString();
+                    final File file = path.toFile();
+                    if (file.isFile()) {
+                        zip.putNextEntry(new ZipEntry(relative));
+                        try (FileInputStream fis = new FileInputStream(file)) {
+                            zip.write(new BytesOf(new InputOf(fis)).asBytes());
+                        }
+                    } else {
+                        final Iterator<Path> paths = new Skipped<>(
+                            1,
+                            new Directory(file).iterator()
+                        );
+                        if (!paths.hasNext()) {
+                            zip.putNextEntry(
+                                new ZipEntry(
+                                    new Joined("/", relative, "").asString()
+                                )
+                            );
+                        }
```

Comment:
@fanifieiev one last thing: what is the purpose of this introduced else branch here? it wasn't there before and it doesn't seem to solve the problem from the original ticket you are solving.

Reply:
victornoel this else branch adds empty folder to zip file, if empty folder or subfolder is present.

Reply:
@fanifieiev it seems also complex for what it does, as a REV I'm not happy with it and I don't want to spend more time explaining what bothers me and what should be done.
So I think it would be safer to remove this and add a todo for this and let the work continue in another ticket/PR.